### ISSUE TITLE
Timeout on attempt to load site.dev

### ISSUE THREAD

**Issue by olivierlacan** [puma/puma-dev#23]
Is there a way to debug timeout issues?

I symlinked a domain that works with Pow and it never managed to load. After trying `-cleanup`, `-setup`, and `-pow` I eventually got to: 

> It works! 

But not much else. I'm just curious how I can observe puma-dev listening somewhere to see if I'm actually hitting it. That's a thing I commonly do with pow and [Powder](https://github.com/rodreegez/powder).

**Comment by evanphx** [comment]
Not sure what you mean you got to "It works!". Did the app load? There are logs at `~/Library/Logs/puma-dev.log`. Anything in there?

**Comment by paulca** [comment]
I'm getting random timeouts too. Here's one backtrace:

```
RuntimeError: There is already a server bound to: /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-5366.sock
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/binder.rb:361:in `add_unix_listener'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/binder.rb:138:in `block in parse'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/binder.rb:85:in `each'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/binder.rb:85:in `parse'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/runner.rb:129:in `load_and_bind'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/single.rb:84:in `run'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/launcher.rb:172:in `run'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/lib/puma/cli.rb:74:in `run'
  /usr/local/var/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/puma-3.5.2/bin/puma:10:in `<top (required)>'
  /usr/local/var/rbenv/versions/2.3.0/bin/puma:23:in `load'
  /usr/local/var/rbenv/versions/2.3.0/bin/puma:23:in `<top (required)>'
```

Here's another one:

```
2016/07/30 12:05:23 http2: panic serving 127.0.0.1:63871: runtime error: slice bounds out of range
goroutine 4703 [running]:
net/http.(*http2serverConn).runHandler.func1(0xc82031ef6f, 0xc820090780, 0xc820028030)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:4050 +0xbc
panic(0x4438e00, 0xc820010040)
	/usr/local/Cellar/go/1.6.3/libexec/src/runtime/panic.go:443 +0x4e9
puma/dev.(*HTTPServer).director(0xc82010fc20, 0xc820488000, 0x0, 0x0)
	/Users/evan/git/puma-dev/src/puma/dev/http.go:69 +0x193
puma/dev.(*HTTPServer).(puma/dev.director)-fm(0xc820488000, 0x0, 0x0)
	/Users/evan/git/puma-dev/src/puma/dev/http.go:79 +0x38
puma/httputil.(*ReverseProxy).ServeHTTP(0xc8201134f0, 0x5a000f0, 0xc820028030, 0xc820486000)
	/Users/evan/git/puma-dev/src/puma/httputil/reverseproxy.go:182 +0x4cb
net/http.serverHandler.ServeHTTP(0xc820105580, 0x5a000f0, 0xc820028030, 0xc820486000)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/server.go:2081 +0x19e
net/http.initNPNRequest.ServeHTTP(0xc8202e8000, 0xc820105580, 0x5a000f0, 0xc820028030, 0xc820486000)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/server.go:2489 +0x221
net/http.(*initNPNRequest).ServeHTTP(0xc82013f070, 0x5a000f0, 0xc820028030, 0xc820486000)
	<autogenerated>:253 +0xb6
net/http.(Handler).ServeHTTP-fm(0x5a000f0, 0xc820028030, 0xc820486000)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:3847 +0x50
net/http.(*http2serverConn).runHandler(0xc820090780, 0xc820028030, 0xc820486000, 0xc8200fa180)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:4060 +0x9f
created by net/http.(*http2serverConn).processHeaderBlockFragment
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:3853 +0x55e
2016/07/30 12:05:23 http2: panic serving 127.0.0.1:63871: runtime error: slice bounds out of range
goroutine 4680 [running]:
net/http.(*http2serverConn).runHandler.func1(0xc82014af6f, 0xc820090780, 0xc82015c0d8)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:4050 +0xbc
panic(0x4438e00, 0xc820010040)
	/usr/local/Cellar/go/1.6.3/libexec/src/runtime/panic.go:443 +0x4e9
puma/dev.(*HTTPServer).director(0xc82010fc20, 0xc820188fc0, 0x0, 0x0)
	/Users/evan/git/puma-dev/src/puma/dev/http.go:69 +0x193
puma/dev.(*HTTPServer).(puma/dev.director)-fm(0xc820188fc0, 0x0, 0x0)
	/Users/evan/git/puma-dev/src/puma/dev/http.go:79 +0x38
puma/httputil.(*ReverseProxy).ServeHTTP(0xc8201134f0, 0x5a000f0, 0xc82015c0d8, 0xc820188ee0)
	/Users/evan/git/puma-dev/src/puma/httputil/reverseproxy.go:182 +0x4cb
net/http.serverHandler.ServeHTTP(0xc820105580, 0x5a000f0, 0xc82015c0d8, 0xc820188ee0)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/server.go:2081 +0x19e
net/http.initNPNRequest.ServeHTTP(0xc8202e8000, 0xc820105580, 0x5a000f0, 0xc82015c0d8, 0xc820188ee0)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/server.go:2489 +0x221
net/http.(*initNPNRequest).ServeHTTP(0xc82013f070, 0x5a000f0, 0xc82015c0d8, 0xc820188ee0)
	<autogenerated>:253 +0xb6
net/http.(Handler).ServeHTTP-fm(0x5a000f0, 0xc82015c0d8, 0xc820188ee0)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:3847 +0x50
net/http.(*http2serverConn).runHandler(0xc820090780, 0xc82015c0d8, 0xc820188ee0, 0xc82013d540)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:4060 +0x9f
created by net/http.(*http2serverConn).processHeaderBlockFragment
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:3853 +0x55e
2016/07/30 12:05:28 http2: panic serving 127.0.0.1:63871: runtime error: slice bounds out of range
goroutine 4704 [running]:
net/http.(*http2serverConn).runHandler.func1(0xc82031ef6f, 0xc820090780, 0xc820028038)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:4050 +0xbc
panic(0x4438e00, 0xc820010040)
	/usr/local/Cellar/go/1.6.3/libexec/src/runtime/panic.go:443 +0x4e9
puma/dev.(*HTTPServer).director(0xc82010fc20, 0xc8204880e0, 0x0, 0x0)
	/Users/evan/git/puma-dev/src/puma/dev/http.go:69 +0x193
puma/dev.(*HTTPServer).(puma/dev.director)-fm(0xc8204880e0, 0x0, 0x0)
	/Users/evan/git/puma-dev/src/puma/dev/http.go:79 +0x38
puma/httputil.(*ReverseProxy).ServeHTTP(0xc8201134f0, 0x5a000f0, 0xc820028038, 0xc8204860e0)
	/Users/evan/git/puma-dev/src/puma/httputil/reverseproxy.go:182 +0x4cb
net/http.serverHandler.ServeHTTP(0xc820105580, 0x5a000f0, 0xc820028038, 0xc8204860e0)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/server.go:2081 +0x19e
net/http.initNPNRequest.ServeHTTP(0xc8202e8000, 0xc820105580, 0x5a000f0, 0xc820028038, 0xc8204860e0)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/server.go:2489 +0x221
net/http.(*initNPNRequest).ServeHTTP(0xc82013f070, 0x5a000f0, 0xc820028038, 0xc8204860e0)
	<autogenerated>:253 +0xb6
net/http.(Handler).ServeHTTP-fm(0x5a000f0, 0xc820028038, 0xc8204860e0)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:3847 +0x50
net/http.(*http2serverConn).runHandler(0xc820090780, 0xc820028038, 0xc8204860e0, 0xc8200fa1c0)
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:4060 +0x9f
created by net/http.(*http2serverConn).processHeaderBlockFragment
	/usr/local/Cellar/go/1.6.3/libexec/src/net/http/h2_bundle.go:3853 +0x55e
2016/07/30 12:05:33 http2: server: error reading preface from client 127.0.0.1:63872: timeout waiting for client preface
! Booted app 'organizer-tito' on socket /Users/paulcampbell/.puma-dev/organizer-tito/tmp/puma-dev-48269.sock
```

And another:

```
* App 'organizer-tito' shutdown and cleaned up
2016/07/30 09:16:35 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:35 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:37 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:37 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:41 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:41 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:42 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:42 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:44 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:16:44 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:17:11 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
2016/07/30 09:17:11 http: proxy error: dial unix /Users/paulcampbell/.puma-dev/organizer.tito/tmp/puma-dev-48269.sock: connect: no such file or directory
```

Typically when it stops responding I just restart the daemon using `launchctl` and everything works again.

**Comment by evanphx** [comment]
Did that happen after the app had been idle a while? Or were you actively
using it and suddenly that happened.

**Comment by paulca** [comment]
@evanphx Yes, I think it was idle a while.

There have been a few cases where I was actively using it and when loading (in Chrome) the screen would just go white.

It could be unrelated, but when I tried firing up another app that didn't have puma in its Gemfile, it seemed to cause the other app to stop working without a restart.

**Comment by evanphx** [comment]
Ok. There was a bug I fixed this weekend (not yet released) that fixed some idle app handling. I'll also go over some of the handling to make it a bit more robust.

**Comment by paulca** [comment]
👍

**Comment by evanphx** [comment]
This should be fixed in v0.8. Please open a new issue if not!
